File: Sources/lambda/Services/NowPlayingService.swift
import AWSLambdaEvents
import AWSLambdaRuntime
import Foundation
import Logging

actor NowPlayingService {
    private static let nowPlayingCacheKey = "NowPlaying"
    private static let revalidateSeconds = 3

    private let cache: MemoryCache
    private let spotifyApi: SpotifyApi
    private let logger: Logger

    init(cache: MemoryCache, spotifyApi: SpotifyApi, logger: Logger) {
        self.cache = cache
        self.spotifyApi = spotifyApi
        self.logger = logger
    }

    func handleNowPlaying() async throws -> NowPlayingResponse {
        do {
            if let cachedResponse: NowPlayingResponse = await cache.get(Self.nowPlayingCacheKey) {
                logger.info("Returning cached now playing response")
                return cachedResponse
            }

            if !Environment.shouldCallSpotify {
                return NowPlayingResponse(
                    isPlaying: false, maintenance: true, status: 200, album: "",
                    albumImageUrl: "", artist: "", songUrl: "", title: ""
                )
            }

            let nowPlayingResponse = try await spotifyApi.getNowPlaying()

            if nowPlayingResponse?.item == nil {
                logger.info("No song currently playing")

                return NowPlayingResponse(
                    isPlaying: false,
                    maintenance: nil,
                    status: 200,
                    album: "",
                    albumImageUrl: "",
                    artist: "",
                    songUrl: "",
                    title: ""
                )
            }

            guard let item = nowPlayingResponse?.item else {
                throw NowPlayingError.missingItem
            }

            let response = NowPlayingResponse(
                isPlaying: nowPlayingResponse?.isPlaying ?? false,
                maintenance: nil,
                status: 200,
                album: item.album.name,
                albumImageUrl: item.album.images.first?.url ?? "",
                artist: item.artists.map { $0.name }.joined(separator: ", "),
                songUrl: item.externalUrls.spotify,
                title: item.name
            )

            await cache.set(Self.nowPlayingCacheKey, value: response, expiration: 5.0)

            return response
        } catch {
            if let spotifyError = error as? SpotifyServiceError {
                logger.error("Error fetching nowplaying data from spotify: \(spotifyError.description)")
            } else {
                logger.error(
                    "Error fetching nowplaying data from spotify: \(error.localizedDescription)")
            }

            return NowPlayingResponse(
                isPlaying: false,
                maintenance: nil,
                status: 500,
                album: "",
                albumImageUrl: "",
                artist: "",
                songUrl: "",
                title: ""
            )
        }
    }
}

enum NowPlayingError: Error {
    case missingItem
}

File: src/Lho.Lambda/Services/NowPlayingService.cs
using Amazon.Lambda.Core;
using Lho.Lambda.Clients.LastFm;
using Lho.Lambda.Clients.Spotify;
using Lho.Lambda.Models;
using Lho.Lambda.Observability;
using Lho.Lambda.RuntimeConfiguration.Options;
using Lho.Lambda.Utils;

namespace Lho.Lambda.Services;

public class NowPlayingService(
  MemoryCache cache,
  SpotifyApi spotifyApi,
  LastFmApi lastFmApi,
  ILambdaLogger logger,
  FeatureFlagsOptions featureFlags)
{
  private const string LastFmProvider = "lastfm";
  private const string SpotifyProvider = "spotify";
  private const string LastFmCacheKey = "NowPlaying:lastfm";
  private const string SpotifyCacheKey = "NowPlaying:spotify";

  public async Task<NowPlayingResponse> HandleNowPlaying(string provider = LastFmProvider)
  {
    return string.Equals(provider, SpotifyProvider, StringComparison.OrdinalIgnoreCase)
      ? await GetSpotifyNowPlaying()
      : await GetNowPlaying();
  }

  public async Task<NowPlayingResponse> GetNowPlaying()
  {
    return await GetCachedNowPlaying(LastFmCacheKey, LastFmProvider, HandleLastFmNowPlaying);
  }

  public async Task<NowPlayingResponse> GetSpotifyNowPlaying()
  {
    return await GetCachedNowPlaying(SpotifyCacheKey, SpotifyProvider, HandleSpotifyNowPlaying);
  }

  private async Task<NowPlayingResponse> GetCachedNowPlaying(
    string cacheKey,
    string provider,
    Func<Task<NowPlayingResponse>> fetch)
  {
    try
    {
      var cachedResponse = cache.Get<NowPlayingResponse>(cacheKey);
      if (cachedResponse is not null)
      {
        logger.LogLine("Returning cached now playing response");
        return cachedResponse;
      }

      var response = await fetch();

      if (response.Status == 200 && !string.IsNullOrEmpty(response.Title))
      {
        cache.Set(cacheKey, response, TimeSpan.FromSeconds(5));
      }

      return response;
    }
    catch (Exception exception)
    {
      logger.LogLine($"Error fetching now playing data from {provider}: {exception}");
      await SentryTelemetry.CaptureExceptionAsync(exception, logger, new Dictionary<string, string?>
      {
        ["operation"] = "now-playing",
        ["provider"] = provider
      });
      return EmptyResponse(status: 500);
    }
  }

  private async Task<NowPlayingResponse> HandleLastFmNowPlaying()
  {
    var recentTracksResponse = await lastFmApi.GetRecentTracks();
    var track = recentTracksResponse.RecentTracks.Tracks.FirstOrDefault();
    if (track is null || !IsNowPlaying(track))
    {
      logger.LogLine("No song currently playing");
      return EmptyResponse(status: 200);
    }

    return new NowPlayingResponse(
        IsPlaying: true,
        Maintenance: false,
        Status: 200,
        Album: track.Album.Text,
        AlbumImageUrl: track.Images.LastOrDefault(image => !string.IsNullOrEmpty(image.Url))?.Url ?? "",
        Artist: track.Artist.Text,
        SongUrl: track.Url,
        Title: track.Name);
  }

  private async Task<NowPlayingResponse> HandleSpotifyNowPlaying()
  {
    if (!featureFlags.ShouldCallSpotify)
    {
      return EmptyResponse(maintenance: true, status: 200);
    }

    var nowPlayingResponse = await spotifyApi.GetNowPlaying();
    if (nowPlayingResponse?.Item is null || !nowPlayingResponse.IsPlaying)
    {
      logger.LogLine("No song currently playing");
      return EmptyResponse(status: 200);
    }

    var item = nowPlayingResponse.Item;
    return new NowPlayingResponse(
        IsPlaying: nowPlayingResponse.IsPlaying,
        Maintenance: false,
        Status: 200,
        Album: item.Album.Name,
        AlbumImageUrl: item.Album.Images.FirstOrDefault()?.Url ?? "",
        Artist: string.Join(", ", item.Artists.Select(artist => artist.Name)),
        SongUrl: item.ExternalUrls.Spotify,
        Title: item.Name);
  }

  private static bool IsNowPlaying(LastFmTrack track)
  {
    return string.Equals(track.Attributes?.NowPlaying, "true", StringComparison.OrdinalIgnoreCase);
  }

  private static NowPlayingResponse EmptyResponse(int status, bool maintenance = false)
  {
    return new NowPlayingResponse(
        IsPlaying: false,
        Maintenance: maintenance,
        Status: status,
        Album: "",
        AlbumImageUrl: "",
        Artist: "",
        SongUrl: "",
        Title: "");
  }
}
